allure-cucumber4-jvm/src/main/java/io/qameta/allure/cucumber4jvm/AllureCucumber4Jvm.java

```diff
@@ -0,0 +1,339 @@
+    private final ThreadLocal<TestCase> currentTestCase = new ThreadLocal<>();
+
+    private final EventHandler<TestSourceRead> featureStartedHandler = this::handleFeatureStartedHandler;
+    private final EventHandler<TestCaseStarted> caseStartedHandler = this::handleTestCaseStarted;
+    private final EventHandler<TestCaseFinished> caseFinishedHandler = this::handleTestCaseFinished;
+    private final EventHandler<TestStepStarted> stepStartedHandler = this::handleTestStepStarted;
+    private final EventHandler<TestStepFinished> stepFinishedHandler = this::handleTestStepFinished;
+
+    public AllureCucumber4Jvm() {
+        this(Allure.getLifecycle());
+    }
+
+    public AllureCucumber4Jvm(final AllureLifecycle lifecycle) {
+        this.lifecycle = lifecycle;
+    }
+
+    @Override
+    public void setEventPublisher(final EventPublisher publisher) {
+        publisher.registerHandlerFor(TestSourceRead.class, featureStartedHandler);
+
+        publisher.registerHandlerFor(TestCaseStarted.class, caseStartedHandler);
+        publisher.registerHandlerFor(TestCaseFinished.class, caseFinishedHandler);
+
+        publisher.registerHandlerFor(TestStepStarted.class, stepStartedHandler);
+        publisher.registerHandlerFor(TestStepFinished.class, stepFinishedHandler);
+    }
+
+    /*
+    Event Handlers
+     */
+
+    private void handleFeatureStartedHandler(final TestSourceRead event) {
+        testSources.addTestSourceReadEvent(event.uri, event);
+    }
+
+    private void handleTestCaseStarted(final TestCaseStarted event) {
+        currentFeatureFile.set(event.testCase.getUri());
+        currentFeature.set(testSources.getFeature(currentFeatureFile.get()));
+
+        currentTestCase.set(event.testCase);
+
+        final Deque<PickleTag> tags = new LinkedList<>(event.testCase.getTags());
+
+        final LabelBuilder labelBuilder = new LabelBuilder(currentFeature.get(), event.testCase, tags);
+
+        final String name = event.testCase.getName();
+        final String featureName = currentFeature.get().getName();
+
+        final TestResult result = new TestResult()
+                .setUuid(getTestCaseUuid(event.testCase))
+                .setHistoryId(getHistoryId(event.testCase))
+                .setFullName(String.format("%s: %s", featureName, name))
+                .setName(name)
+                .setLabels(labelBuilder.getScenarioLabels())
+                .setLinks(labelBuilder.getScenarioLinks());
+
+        final ScenarioDefinition scenarioDefinition =
+                testSources.getScenarioDefinition(currentFeatureFile.get(), currentTestCase.get().getLine());
+        if (scenarioDefinition instanceof ScenarioOutline) {
+            result.setParameters(
+                    getExamplesAsParameters((ScenarioOutline) scenarioDefinition)
+            );
+        }
+
+        if (currentFeature.get().getDescription() != null && !currentFeature.get().getDescription().isEmpty()) {
```

Comment:
cache current feature value in the beginning of the method, instead of using thread local each time

Reply:
does this makes sense only when get() called multiple times within method or store value to local variable for each method even if it is used once?

Reply:
the problem is that value actually may be changed between calls and you'll probably get unexpected result in this case.

> store value to local variable for each method even if it is used once?

This is not required, but I personally like it (easier to read/refactor the code).

Also note that  `threadLocal.get()` may return null value, so the correct pattern of using it is have a null check or calculate value to variable, set value to thread local, and then use calculated variable

Reply:
Added local variables withing method to cache calls to ThreadLocal variables. But not sure how to make these variable "calculateable" because it is `null` then something went wrong during `TestCaseStarted` started event

Reply:
> Added local variables withing method to cache calls to ThreadLocal variables. But not sure how to make these variable "calculateable" because it is `null` then something went wrong during `TestCaseStarted` event handling

you probably need to skip method execution in this case in order to prevent NPE. But we can keep it this way for now

Reply:
Good idea, maybe adding errors to console log?

Reply:
yep, use slf4 and log such errors

Reply:
Done